feat(extension): Add option to enable async workers in Flask and Django (#1986)
canonical/charmcraft: docs/howto/flask-async.rst

======================================================
How to write a Kubernetes charm for an Async Flask app
======================================================

In this how to we will configure 12 Factor Flask application
to use asynchronous Gunicorn workers to be able to serve
to multiple users easily.

Make the rock async
===================

Before packing the rock make sure to put the following in ``requirements.txt``
file:

.. literalinclude:: code/flask-async/requirements.txt

Configure the async application
===============================

Now let's enable async Gunicorn workers using a configuration option. We will
expect this configuration option to be available in the Flask app configuration
under the keyword ``webserver-worker-class``. Verify that the new configuration
has been added using
``juju config flask-async-app | grep -A 6 webserver-worker-class:`` which should
show the configuration option.

.. note::

    The ``grep`` command extracts a portion of the configuration to make

>     it easier to check whether the configuration option has been added.

Reviewer: Nitpick: I think you can remove this note. I think it makes more sense in a tutorial versus a how-to guide.

The worker class can be changed using Juju:

.. literalinclude:: code/flask-async/task.yaml
    :language: bash
    :start-after: [docs:config-async]
    :end-before: [docs:config-async-end]
    :dedent: 2

Now you can run
``curl --parallel --parallel-immediate --resolve flask-async-app:80:127.0.0.1 \
http://flask-async-app/io http://flask-async-app/io http://flask-async-app/io \
http://flask-async-app/io http://flask-async-app/io``

> in they will all return at the same time.

Reviewer: For my understanding, each instance of `http://flask-async-app/io` corresponds to a different async runner? Does "they will all return at the same time" mean that the async runners all respond to the `curl` request at the same time?

Perhaps it would be helpful to include example output here, although I don't know how universal the output is.

======================================================
How to write a Kubernetes charm for an Async Flask app
======================================================

In this how-to guide we will configure a 12-factor Flask
application to use asynchronous Gunicorn workers to be
able to serve to multiple users easily.

Make the rock async
===================

Before packing the rock make sure to put the following in ``requirements.txt``
file:

.. literalinclude:: code/flask-async/requirements.txt

> Configure the async application

Reviewer 2: @erinecon should we say that they should repack and deploy the charm with the new rock here?

Reviewer: Yes, that's a good point. I think it could be a part of the `Make the rock async` section. Something like:

> Pack the rock using `rockcraft pack` and redeploy the charm with the new rock using `juju refresh`. 

If there's any concern about being too vague with the `juju refresh` command, then we could link to the Juju docs.

===============================

Now let's enable async Gunicorn workers using a configuration option. We will
expect this configuration option to be available in the Flask app configuration
under the keyword ``webserver-worker-class``. Verify that the new configuration
has been added using
``juju config flask-async-app | grep -A 6 webserver-worker-class:`` which should
show the configuration option.

The worker class can be changed using Juju:

.. literalinclude:: code/flask-async/task.yaml
    :language: bash
    :start-after: [docs:config-async]
    :end-before: [docs:config-async-end]
    :dedent: 2

Now you can run
``curl --parallel --parallel-immediate --resolve flask-async-app:80:127.0.0.1 \
http://flask-async-app/io http://flask-async-app/io http://flask-async-app/io \
http://flask-async-app/io http://flask-async-app/io``

> in they will all return at the same time.

Reviewer 2: Perhaps review this, the `in` doesn't make sense? Maybe you meant `and`?

Output will be similar to following:

.. code-block:: bash

   ok
   ok
   ok
   ok
   ok

.. note::

>     It might take a short time for the configuration to take effect.

Reviewer 2: Perhaps include a note how to tell when it is done? I think the charm should be back in active?

> ======================================================
> How to write a Kubernetes charm for an Async Flask app

Reviewer 3: ```suggestion
.. _write-a-kubernetes-charm-for-an-async-flask-app:

How to write a Kubernetes charm for an async Flask app
```

======================================================

> In this how-to guide we will configure a 12-factor Flask

Reviewer 3: The use of "we" suggests a tutorial, whereas in a how-to we should orient around "you" instead.

application to use asynchronous Gunicorn workers to be
able to serve to multiple users easily.

Make the rock async
===================

> Before packing the rock make sure to put the following in ``requirements.txt``

Reviewer 3: ```suggestion
Before packing the rock, make sure to put the following in its ``requirements.txt``
```

file:

.. literalinclude:: code/flask-async/requirements.txt

Configure the async application

> ===============================

Reviewer 3: ```suggestion
-------------------------------
```

> Now let's enable async Gunicorn workers using a configuration option. We will

Reviewer 3: ```suggestion
Now let's enable async Gunicorn workers. We will
```

expect this configuration option to be available in the Flask app configuration

> under the keyword ``webserver-worker-class``. Verify that the new configuration

Reviewer 3: ```suggestion
under the ``webserver-worker-class`` key. Verify that the new configuration
```

> has been added using
> ``juju config flask-async-app | grep -A 6 webserver-worker-class:`` which should
> show the configuration option.

Reviewer 3: ```suggestion
has been added by running:

.. code:: bash

  juju config flask-async-app | grep -A 6 webserver-worker-class:

The result should contain the key.
```

The worker class can be changed using Juju:

.. literalinclude:: code/flask-async/task.yaml
    :language: bash
    :start-after: [docs:config-async]
    :end-before: [docs:config-async-end]
    :dedent: 2

> 

Reviewer 3: Change the duration as needed.

```suggestion

It can take up to a minute for the configuration to take effect. When the
configuration changes, the charm will re-enter the active state.

```

> Now you can run
> ``curl --parallel --parallel-immediate --resolve flask-async-app:80:127.0.0.1 \
> http://flask-async-app/io http://flask-async-app/io http://flask-async-app/io \
> http://flask-async-app/io http://flask-async-app/io``
> and they will all return at the same time.

Reviewer 3: ```suggestion
Test that the workers are operating in parallel by sending multiple
simultaneous requests with curl:

.. code:: bash

  curl --parallel --parallel-immediate --resolve flask-async-app:80:127.0.0.1 \
  http://flask-async-app/io http://flask-async-app/io http://flask-async-app/io \
  http://flask-async-app/io http://flask-async-app/io
```

> Output will be similar to following:
> 
> .. code-block:: bash

Reviewer 3: ```suggestion
The results should arrive simultaneously and contain five instances of ``ok``:

.. terminal::
```

   ok
   ok
   ok
   ok
   ok

> .. note::
> 
>     When the configuration is done the charm will be in ``active`` state.
>     It might take a short time for the configuration to take effect.

Reviewer 3: Covered in an earlier suggestion.

```suggestion
```
